### PR TITLE
feat(api): add PUT /certificate/verify

### LINKED ISSUE #51298: PUT /certificate/verify

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #51298

### CODE REVIEW

**Comment by ShaunSHamilton** [comment]
Team, currently, I am stuck with:

```bash
The types of 'response.type' are incompatible between these types.
                Type '"success"' is not assignable to type '"info"'.ts(2345)
```

I cannot figure out why TS is saying this, when there is a union of two distinct types.

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [inline api/src/routes/certificate.ts:245]: To discuss:

`api-server/` uses Loopback to template this. Do we need/want to do that in the new api?

I do not see the need, and do not want to add the complexity.

Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:245]: Discussed in meeting: As long as functionality is the same, this is fine as is.

**Review by moT01 — COMMENTED** [review]
Nice work on this @ShaunSHamilton :+1: It does not look like an easy one.

Everything seems to be working :+1: I found a few things to bring up...

File: api/src/schemas.ts
Comment (ShaunSHamilton) [reply api/src/schemas.ts:575]: A lot of the schemas need work, but we have to do an overhaul of them anyway once we change to returning the responses we actually want. So, I will leave this for another PR(s)

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:170]: No, that is the old Loopback `Email.send$` API. This is our one defined here:
https://github.com/ShaunSHamilton/freeCodeCamp/blob/065c3becd82969884cb6d9080e6dee5604e3a307/api/src/plugins/mailer.ts#L10-L16

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts]: We currently have a bug that prevents this from working the the current API, but I have added that as well.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:68]: I recommend using either `jest.spyOn` or `jest.replaceProperty`, here. Probably the latter, since you don't seem to be testing how the method is called.

That means you can use `jest.restoreAllMocks` in `afterEach` rather than having to restore it inside the test. It keeps the test more focused on what it's trying to do.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:52]: Couple of things.

`response.body` should give you the parsed body, so you don't need `JSON.parse` here. There's a few other cases later on where you can simplify this.

Also, we want to send an error response with `flash.wrong-name`, right? Regardless, a serialization error should never occur, since it means our handler has tried to return something unexpected

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @ShaunSHamilton I'll do a deeper review when I have a chance, but I found a few things that're worth addressing.

File: api/src/utils/env.ts
Comment (ShaunSHamilton) [reply api/src/utils/env.ts:113]: Will do, but do we actually care about that? I see we do that for most of the variables, but in the case of development (test) only variables, `undefined` seems fine when the check is for something explicit like `=== 'true'` (definitely set on purpose).

File: api/src/routes/certificate.test.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.test.ts:281]: No, nothing else. Our `Certification` enum does not include the old full stack cert. Also, I thought we could no longer claim the `LegacyInfoSecQa` cert, but we do still allow it.

File: api/src/utils/env.ts
Comment (ojeytonwilliams) [reply api/src/utils/env.ts:113]: It's probably fine either way, but the argument for including it is that it's one less thing to think about. i.e. nobody need wonder why some variables are asserted and others are not.

File: api/src/routes/certificate.test.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.test.ts:52]: This is `response`:

```bash
{"header": {"access-control-allow-credentials": "true", "access-control-allow-headers": "Origin, X-Requested-With, Content-Type, Accept, Csrf-Token", "access-control-allow-methods": "GET, PUT, POST, DELETE", "access-control-allow-origin": "https://www.freecodecamp.org", "cache-control": "no-store", "connection": "close", "content-length": "643", "content-security-policy": "frame-ancestors 'none'", "content-type": "application/json; charset=utf-8", "date": "Thu, 12 Oct 2023 11:50:17 GMT", "set-cookie": ["csrf_token=PR8jxEZp-eEgwvv3I8KblTOhZzLBBApKvhv1FMA8RE6UmGUJ6d0Y; Domain=localhost; Path=/; SameSite=Strict"], "vary": "Origin, Accept-Encoding", "x-content-type-options": "nosniff", "x-frame-options": "DENY"}, "req": {"data": {"certSlug": "responsive-web-design"}, "headers": {"content-type": "application/json", "cookie": ["_csrf=n5tHCOG-NNVDfVlbuWgF-B44; Path=/; HttpOnly; SameSite=Strict", "csrf_token=4ND3axq1-SCLBQoGxcfDIXfqNo0NqQEKdvEE99F-nSI6vexDdjM8; Domain=localhost; Path=/; SameSite=Strict", "sessionId=jdIUR6yfm4hrrPk_2rAM-A_7mnQ0i5zB.BDi1lbSZGPdHXVayT21SL6qOXDwLdHPcscETkbFxCFg; Path=/; Expires=Thu, 12 Oct 2023 12:50:17 GMT; HttpOnly"], "csrf-token": "4ND3axq1-SCLBQoGxcfDIXfqNo0NqQEKdvEE99F-nSI6vexDdjM8", "origin": "https://www.freecodecamp.org"}, "method": "PUT", "url": "http://127.0.0.1:44511/certificate/verify"}, "status": 400, "text": "{\"response\":{\"type\":\"info\",\"message\":\"flash.incomplete-steps\",\"variables\":{\"name\":\"Responsive Web Design\"}},\"isCertMap\":{\"isRespWebDesignCert\":false,\"isJsAlgoDataStructCert\":false,\"isFrontEndLibsCert\":false,\"is2018DataVisCert\":false,\"isApisMicroservicesCert\":false,\"isInfosecQaCert\":false,\"isQaCertV7\":false,\"isInfosecCertV7\":false,\"isFrontEndCert\":false,\"isBackEndCert\":false,\"isDataVisCert\":false,\"isFullStackCert\":false,\"isSciCompPyCertV7\":false,\"isDataAnalysisPyCertV7\":false,\"isMachineLearningPyCertV7\":false,\"isRelationalDatabaseCertV8\":false,\"isCollegeAlgebraPyCertV8\":false,\"isFoundationalCSharpCertV8\":false},\"completedChallenges\":[]}"}
```

There is no `response.body`.

Comment (ShaunSHamilton) [reply api/src/routes/certificate.test.ts]: EDIT (_without editing_): Turns out there is a `body`, it is just a getter.

**Comment by moT01** [comment]
Want to take another look at this when you get a chance @ojeytonwilliams

**Review by moT01 — APPROVED** [review]

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:50]:
```suggestion
      test('should return 400 if no certSlug', async () => {
```

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:106]: Could you add a test for the expected body?

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:71]: I think it's worth testing the response body here, too.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:54]:
```suggestion
        }).send({});
```
Same thing, but shorter.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:294]:
```suggestion
        const unclaimableCerts = ['fake-slug'];
```
So we always test the branch where the slug is invalid.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
This was a tough endpoint. Nice job, @ShaunSHamilton 

I have a few suggestions, but the functionality seems solid.

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:180]: I disagree, just because we definitely want the DB to be our source of truth for this. So, I would prefer keeping the use of `updatedUser`.

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:202]: Possibly, but the `updatedUser` is the most-likely to be correct. Theoretically, an old db op could have been performed between the initial `user` read, and the `updatedUser` read. So, seeing as we have a _more up-to-date_ record, I would like to use it.

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:353]: Thank you, but I do not understand the syntax here. What does `[number]` do?

File: api/src/routes/certificate.ts
Comment (ojeytonwilliams) [reply api/src/routes/certificate.ts:202]: In principle, sure, but does that matter? Is there a meaningful difference between the db op happening in the milliseconds between these calls and a few milliseconds after?

File: api/src/routes/certificate.ts
Comment (ojeytonwilliams) [reply api/src/routes/certificate.ts]: It gives you the type of the array elements. E.g. if you have `type Names = string[]` then `Names[number]` gives you `string`

Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts]: Ah. That is confusing.

File: api/src/routes/certificate.ts
Comment (ShaunSHamilton) [reply api/src/routes/certificate.ts:228]: Yes, because a Camper _could_ have `completedChallenges` | `username` | `email` change which would break the following logic.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:305]: This leaves a fair chunk of the body uncovered. Ideally we want to know what `completedChallenges` and `isCertMap` are.

Since the client mindlessly updates the user with all the properties, it would be good to know what they are.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Sorry for the delay, Shaun. I think we're nearly there :crossed_fingers:

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:311]: Should this be empty?

Comment (ShaunSHamilton) [reply api/src/routes/certificate.test.ts:311]: No, probably not.

File: api/src/routes/certificate.test.ts
Comment (ojeytonwilliams) [inline api/src/routes/certificate.test.ts:399]: Conditional testing is a bit odd, tbh. I'd just test both or one of them (without the conditional).

Please merge regardless, I just wanted to make the observation!

**Review by ojeytonwilliams — APPROVED** [review]
Since I can't dismiss my review, I'll approve so I'm not in the way. It would be good if we had another pair of eyes on this before merging, though.

**Review by Sembauke — APPROVED** [review]

**Review by Sembauke — APPROVED** [review]

**Pull request merged**
